### PR TITLE
Generate schedule upon user creation

### PR DESCRIPTION
**Pull request by mpgxvii**
- Currently, when a new user is added, we would need to wait for the task that generates the schedule every hour to run before a new user gets a schedule. This would generate the schedule immediately when a user is added.

### CODE REVIEW

**Review by yatharthranjan — APPROVED** [review]
LGTM, a minor comment

File: src/main/java/org/radarbase/appserver/controller/RadarUserController.java [hunk 1]
```diff
@@ -91,6 +95,11 @@ public ResponseEntity<FcmUserDto> addUserToProject(
           null)) {
         if (forceFcmToken) this.userService.checkFcmTokenExistsAndReplace(userDto);
         FcmUserDto user = this.userService.saveUserInProject(userDto);
+        if (user != null) {
+          // Generate schedule for user upon saving
+          this.scheduleService.generateScheduleUsingProjectIdAndSubjectId(projectId, user.getSubjectId());
```
[the remark this answers is not in the document]
Comment (mpgxvii) [reply src/main/java/org/radarbase/appserver/controller/RadarUserController.java]: I've added a `UserListener` instead to generate the schedule upon adding the user entity. Is this ok?

File: src/main/java/org/radarbase/appserver/service/questionnaire/protocol/GithubProtocolFetcherStrategy.java [hunk 2]
```diff
@@ -259,17 +266,11 @@ private Map<String, URI> getProtocolDirectories() throws IOException {
...
-            GithubContent content = mapper.readValue(uri.toURL(), GithubContent.class);
-            return mapper.readValue(content.getContent(), Protocol.class);
-        } else {
-            log.error("Error getting Protocol from URL {} : {}", uri.toString(), response);
-            throw new ResponseStatusException(
-                    HttpStatus.valueOf(response.statusCode()), "Protocol could not be retrieved");
-        }
+        String contentString = this.githubClient.getGithubContent(uri.toString());
+        final ObjectMapper mapper =
+                objectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
```
Comment (yatharthranjan) [inline src/main/java/org/radarbase/appserver/service/questionnaire/protocol/GithubProtocolFetcherStrategy.java:271]: Would this not change the behaviour of the injected mapper (and could have unintended consequence in other places that it is used)?

I would create a new ObjectMapper at the top of the class as copy of this one like

```
private final ObjectMapper localMapper;
```

Then init it in the constructor - 
```
this.localMapper = new ObjectMapper(objectMapper);
this.localMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
```

File: src/main/java/org/radarbase/appserver/service/UserService.java [hunk 3]
```diff
@@ -164,7 +169,10 @@ public FcmUserDto saveUserInProject(FcmUserDto userDto) {
       User newUser = userConverter.dtoToEntity(userDto).setProject(project.get());
       // maintain a bi-directional relationship
       newUser.getUsermetrics().setUser(newUser);
-      return userConverter.entityToDto(this.userRepository.save(newUser));
+      User savedUser = this.userRepository.save(newUser);
+      // Generate schedule for user
+      this.scheduleService.generateScheduleUsingProjectIdAndSubjectId(savedUser.getProject().getProjectId(), savedUser.getSubjectId());
```
Comment (yatharthranjan) [inline src/main/java/org/radarbase/appserver/service/UserService.java:174]: Do we need to do this in the update user method too (for example if attribute, timezone or enrolment date is changed)?

Comment (mpgxvii) [reply src/main/java/org/radarbase/appserver/service/UserService.java]: Yup made the changes.

**Pull request merged**